bids/reports/report.py

```diff
@@ -0,0 +1,115 @@
+        if config is None:
+            root = dirname(abspath(__file__))
+            config = pathjoin(root, 'config', 'converters.json')
+            with open(config) as fobj:
+                config = json.load(fobj)
+
+        if not isinstance(config, dict):
+            raise ValueError('Input config must be None, dict, or path to json '
+                             'file containing dict.')
+
+        self.config = config
+
+    def generate(self, task_converter=None):
+        """Generate the methods section.
+
+        Parameters
+        ----------
+        task_converter : :obj:`dict`
+            A dictionary with information for converting task names from BIDS
+            filename format to human readable strings.
+        """
+        if task_converter is None:
+            task_converter = {'rest': 'resting state'}
+
+        self.config['task'] = task_converter
+
+        descriptions = []
+
+        subjs = self.layout.get_subjects()
+        for sid in subjs:
+            description = self._report(subj=sid)
+            descriptions.append(description)
+        counter = Counter(descriptions)
+        print('Number of patterns detected: {0}'.format(len(counter.keys())))
+        print(utils.warnings())
+        return counter
+
+    def _report(self, subj):
+        """Write a report.
+
+        Parameters
+        ----------
+        subj : :obj:`str`
+            Subject ID.
+
+        Attributes
+        ----------
+        layout : :obj:`bids.grabbids.BIDSLayout`
+            Layout object for a BIDS dataset.
+        config : :obj:`dict`
+            Configuration info for methods generation.
+
+        Returns
+        -------
+        description : :obj:`str`
+            A publication-ready report of the dataset's data acquisition
+            information. Each scan type is given its own paragraph.
+        """
+        description_list = []
+        sessions = self.layout.get_sessions()
+        if sessions:
+            for ses in sessions:
+                niftis = self.layout.get(subject=subj, session=ses, extensions='nii.gz')
+
+                if niftis:
+                    description_list.append('For session {0}:'.format(ses))
+                    description_list += utils.parse_niftis(self.layout, niftis, subj,
+                                                           ses, self.config)
+                    metadata = self.layout.get_metadata(niftis[0].filename)
+        else:
+            niftis = self.layout.get(subject=subj, extensions='nii.gz')
+
+            if niftis:
+                description_list += utils.parse_niftis(self.layout, niftis, subj, None, self.config)
+                metadata = self.layout.get_metadata(niftis[0].filename)
+            else:
+                raise Exception('No niftis for subject {0}'.format(subj))
+
+        # Assume all data were converted the same way and use the first nifti file's
+        # json for conversion information.
+        if 'metadata' not in vars():
```

Comment:
If this is just the final paragraph? Is it essential that we fail? I think if the user can get back *some* usable information, it might be worth issuing a `warning` instead, and still returning `description`. But I haven't thought about it, so maybe this doesn't make sense.

Reply:
This should raise an exception if no metadata could be found for any of the scans in the dataset. At least for now, if that exception is raised then description is empty anyway.

When there are only niftis, the only information that could be extracted are the number of slices, matrix size, field of view, and voxel size. At least at the moment, the parsing functions won't return that information if metadata couldn't be found, so nothing would be returned. If we end up making those functions more flexible, then I guess very short descriptions could be created and returned for the scans, but I don't know if people will use the module in cases where so little information can be extracted.

Reply:
Ah, that makes sense.